File: javascript/systemparams/machine.js
const FrameEvent = require("./framelang/framelang");

// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

class SystemParams {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    
    constructor (stateMsg,enterMsg) {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sBegin_;
        this.#compartment = new SystemParamsCompartment(this.#state);
        this.#nextCompartment = null;
        this.#compartment.StateArgs["stateMsg"] = stateMsg;
        this.#compartment.EnterArgs["enterMsg"] = enterMsg;
        
        // Initialize domain
        
        // Send system start event
        const frameEvent = FrameEvent(">", this.#compartment.EnterArgs);
        this.#mux(frameEvent);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sBegin_:
                this.#sBegin_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sBegin_(e) {
        switch (e._message) {
            case ">":
                {
                this.print_do((this.#compartment.StateArgs["stateMsg"]) + " " + (e._parameters["enterMsg"]));
                return;
                }
                
        }
    }
    
    //===================== Actions Block ===================//
    
    print_do (msg) {
        
        console.log(msg)
    
    }
    
    // Unimplemented Actions
    
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    
    
};


//=============== Compartment ==============//

class SystemParamsCompartment {

  constructor(state) {
      this.state = state
  }
  
  StateArgs = {};
  StateVars = {};
  EnterArgs = {};
  ExitArgs = {};
  _forwardEvent = FrameEvent.call(this)
}




class SystemParamsController extends SystemParams {

constructor(stateMsg,enterMsg) {
  super(stateMsg,enterMsg)
}
};


module.exports = SystemParamsController

File: python/systemparams/machine.py
# emitted from framec_v0.10.0
# get include files at https://github.com/frame-lang/frame-ancillary-files
from framelang.framelang import FrameEvent

class SystemParams:
    
    def __init__(self, stateMsg: str,enterMsg: str):
        
        # Create and intialize start state compartment.
        self.__state = self.__systemparams_state_Begin
        self.__compartment: 'SystemParamsCompartment' = SystemParamsCompartment(self.__state)
        self.__next_compartment: 'SystemParamsCompartment' = None
        self.__compartment.state_args["stateMsg"] = stateMsg
        self.__compartment.enter_args["enterMsg"] = enterMsg
        
        # Initialize domain
        
        # Send system start event
        frame_event = FrameEvent(">", self.__compartment.enter_args)
        self.__mux(frame_event)
    
    # ====================== Multiplexer ==================== #
    
    def __mux(self, e):
        if self.__compartment.state == self.__systemparams_state_Begin:
            self.__systemparams_state_Begin(e)
        
        if self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            if(next_compartment.forward_event is not None and 
               next_compartment.forward_event._message == ">"):
                self.__mux(FrameEvent( "<", self.__compartment.exit_args))
                self.__compartment = next_compartment
                self.__mux(next_compartment.forward_event)
            else:
                self.__do_transition(next_compartment)
                if next_compartment.forward_event is not None:
                    self.__mux(next_compartment.forward_event)
            next_compartment.forward_event = None
    
    # ===================== Machine Block =================== #
    
    def __systemparams_state_Begin(self, e):
        if e._message == ">":
            self.print_do((self.__compartment.state_args["stateMsg"]) + " " + e._parameters["enterMsg"])
            return
        
    
    # ===================== Actions Block =================== #
    
    
    def print_do(self,msg: str):
        
        print(msg)
    
    
    # Unimplemented Actions
    
    
    # =============== Machinery and Mechanisms ============== #
    
    def __transition(self, compartment: 'SystemParamsCompartment'):
        self.__next_compartment = compartment
    
    def  __do_transition(self, next_compartment: 'SystemParamsCompartment'):
        self.__mux(FrameEvent("<", self.__compartment.exit_args))
        self.__compartment = next_compartment
        self.__mux(FrameEvent(">", self.__compartment.enter_args))
    

# ===================== Compartment =================== #

class SystemParamsCompartment:

    def __init__(self, state):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = FrameEvent(None, None)
    


# ********************

#class SystemParamsController(SystemParams):
	#def __init__(self,stateMsg: str,enterMsg: str):
	    #super().__init__(stateMsg: str,enterMsg: str)

# ********************


